interop-testing/src/main/java/io/grpc/testing/integration/TestServiceImpl.java

```diff
@@ -217,6 +232,9 @@ public void onNext(StreamingOutputCallRequest request) {
 
       @Override
       public void onCompleted() {
+        if (oobTestLocked) {
+          lock.release();
```

Comment:
The `onError()` discussion above made me realize neither of these releases may be called. We should instead use ServerCallStreamObserver's cancel and close handlers.

```java
  @Override
  public StreamObserver<Messages.StreamingOutputCallRequest> fullDuplexCall(
      StreamObserver<Messages.StreamingOutputCallResponse> plainResponseObserver) {
    final ServerCallStreamObserver responseObserver = (ServerCallStreamObserver) plainResponseObserver;
    class MyStreamObserver implements StreamObserver<StreamingOutputCallRequest> {
      ...
      void cleanUp() {
        if (oobTestLocked) {
          lock.release();
        }
      }
    }

    MyStreamObserver mso = new MyStreamObserver();
    responseObserver.setOnCancelHandler(mso::cleanUp);
    responseObserver.setOnCloseHandler(mso::cleanUp);
    return mso;
```

Reply:
Because of a StackOverflow question I was answering today, I think I was wrong about this. We do guarantee to call the normal onError/onCompleted, it is just you can't tell cancellations after that point. So either approach works. Sorry.